### PR TITLE
Optimize the speed of creating disks

### PR DESCRIPTION
**Pull request by swimfish09**
[https://github.com/ceph/ceph-iscsi-cli/issues/143](url)

before:
image
now:
image

### CODE REVIEW

**Comment by swimfish09** [comment]
@mikechristie

File: ceph_iscsi_config/gateway.py [hunk 1]
```diff
@@ -412,14 +412,21 @@ def map_luns(self, config):
...
+                tpg_stg_object_dict[str(tpg.tag) + '-' + l.storage_object.name] = l.storage_object.name
+        
         # process each storage object added to the gateway, and map to the tpg
         for stg_object in target_stg_object:
 
             for tpg in self.tpg_list:
                 self.logger.debug("processing tpg{}".format(tpg.tag))
 
-                if not self.lun_mapped(tpg, stg_object):
+                if not tpg_stg_object_dict.has_key(str(tpg.tag) + "-" + stg_object.name):
```
Comment (mikechristie) [inline ceph_iscsi_config/gateway.py]: Did you see my comment:

https://github.com/ceph/ceph-iscsi-config/pull/100#discussion_r250003141

LUN() does a exists check for us, so there is no need to create the dict. Did my suggestion in that comment not perform as well as the dict based approach above?

Comment (swimfish09) [reply ceph_iscsi_config/gateway.py]: yeah, i see. but when catch a exception, configuration will be cleaned and the program exit.

Comment (swimfish09) [reply ceph_iscsi_config/gateway.py]: image

**Comment by dillaman** [comment]
@swimfish09 Can you address the tox errors?

File: ceph_iscsi_config/gateway.py (hunk 1 above)
Comment (mikechristie) [reply ceph_iscsi_config/gateway.py]: It's not clear to me how you hit that. Won't the code:

```
        except RTSLibError as err:
            if "already exists in configFS" not in str(err):
                self.logger.error("LUN mapping failed: {}".format(err))
                self.error = True
                self.error_msg = str(err)
                return

           #Already created. Ignore and loop to the next tpg.
           continue
```

Catch the exception so it does not get escalated up to exception_handler()?

Comment (swimfish09) [reply ceph_iscsi_config/gateway.py]: use this code, rbd-target-gw.log with no error 
image

**Comment by swimfish09** [comment]
> @swimfish09 Can you address the tox errors?
rbd-target-gw.log with no error

**Comment by dillaman** [comment]
@swimfish09 These errors from GitHub [1]

[1] https://jenkins.ceph.com/job/ceph-iscsi-tox/49/console

**Comment by swimfish09** [comment]
Add a dict to check lun is or not mapped will cause errors, the same lun will be mapped twice 
image

When i add code print self.tpg_list to method map_luns, i found self.tpg_list contains repeat tpg. 
Is this correct that self.tpg_list contains repeat tpg?
image
image

**Comment by swimfish09** [comment]
Is this is a error? tpg_list contains repeat tpg
image

image

**Comment by dillaman** [comment]
@mikechristie look good to you?

**Comment by mikechristie** [comment]
I am checking my own review comment to see why it does not work.

**Comment by mikechristie** [comment]
For some reason you cannot attach a file in a nested comment, so I am replying here.

Attached is a patch that works for me and avoids the building of the tpg/lun cache. Exceptions are handled as expected, so if you are hitting issues remove the generic handling in rbd-target-api in  unhandled_exception(), so we can get a trace and double check it is coming from this path.

[skipscan.txt](https://github.com/ceph/ceph-iscsi/files/2882426/skipscan.txt)

**Comment by mikechristie** [comment]
Doh. I posted that then realized for disk creation we do not even need to loop over all the storage_objects too since we just want to add that one disk. Let me open a new PR, because we also do not need to do lun.map_lun->define_luns too (we just need to map the one lun).

**Comment by mikechristie** [comment]
I integrated the idea of your patch to remove the excess lun dir scans into my patchset here:

https://github.com/ceph/ceph-iscsi/pull/34

but based it on my patches. I'm gonna close this PR as a result.

**Pull request closed without merging**